mlir/lib/Quantum/Transforms/PropagateSimpleStatesAnalysis.hpp

```diff
@@ -0,0 +1,264 @@
+         {"PauliX", QubitState::ZERO},
+         {"PauliY", QubitState::ZERO},
+         {"PauliZ", QubitState::ONE},
+     }},
+
+    {QubitState::PLUS,
+     {
+         {"Hadamard", QubitState::ZERO},
+         {"PauliX", QubitState::PLUS},
+         {"PauliY", QubitState::MINUS},
+         {"PauliZ", QubitState::MINUS},
+         {"S", QubitState::LEFT},
+     }},
+
+    {QubitState::MINUS,
+     {
+         {"Hadamard", QubitState::ONE},
+         {"PauliX", QubitState::MINUS},
+         {"PauliY", QubitState::PLUS},
+         {"PauliZ", QubitState::PLUS},
+         {"S", QubitState::RIGHT},
+     }},
+
+    {QubitState::LEFT,
+     {
+         {"Hadamard", QubitState::RIGHT},
+         {"PauliX", QubitState::RIGHT},
+         {"PauliY", QubitState::LEFT},
+         {"PauliZ", QubitState::RIGHT},
+         // We leave in S+ to indicate the FSM structure
+         // The actual implementation is `quantum.custom "S"() %in {adjoint}`
+         //{"S+", QubitState::PLUS},
+     }},
+
+    {QubitState::RIGHT,
+     {
+         {"Hadamard", QubitState::LEFT},
+         {"PauliX", QubitState::LEFT},
+         {"PauliY", QubitState::RIGHT},
+         {"PauliZ", QubitState::LEFT},
+         // We leave in S+ to indicate the FSM structure
+         // The actual implementation is `quantum.custom "S"() %in {adjoint}`
+         //{"S+", QubitState::MINUS},
+     }},
+};
+
+class PropagateSimpleStatesAnalysis {
+  public:
+    PropagateSimpleStatesAnalysis(Operation *target)
+    {
+        // `target` is a qnode function
+        // We restrict the analysis to gates at the top-level body of the function
+        // This is so that gates inside nested regions, like control flows, are not valid targets
+        // e.g.
+        //   func.func circuit() {
+        //      %0 = |0>
+        //      %1 = scf.if (condition) {
+        //          true:  Hadamard
+        //          false: PauliZ
+        //      }
+        //   }
+        // since depending on the control flow path, %1 can be in multiple different states
+
+        // Two kinds of operations produce qubit values: extract ops and custom ops
+        // For extract ops, if its operand is a alloc op directly, then it's a starting qubit and is
+        // in |0>. Then the FSM propagates the states through the custom op gates
+
+        target->walk([&](quantum::ExtractOp op) {
+            // Do not analyze any operations in invalid ops or regions.
+            // The only valid ops are the custom/extract ops whose immediate parent is the qnode
+            // function. With this, we skip over regions like control flow.
+            if (!isImmediateChild(op, target)) {
+                return;
+            }
+
+            // starting qubits are in |0>
+            // We restrict "starting qubits" to those who are extracted immediately from alloc ops
+            if (isa<quantum::AllocOp>(op->getOperand(0).getDefiningOp())) {
+                qubitValues[op->getResult(0)] = QubitState::ZERO;
```

Comment:
Minor, but since we have the concrete op type we can make the code more expressive
```suggestion
            if (isa<quantum::AllocOp>(op->getQreg().getDefiningOp())) {
                qubitValues[op->getQubit()] = QubitState::ZERO;
```

Reply:
Nice! https://github.com/PennyLaneAI/catalyst/pull/1154/commits/4943784e92c08666fe9b1fed96908de5e24aba0c